src/base/BaseV4Quoter.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: UNLICENSED
+
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {QuoterRevert} from "../libraries/QuoterRevert.sol";
+import {SqrtPriceLimitHelper} from "../libraries/SqrtPriceLimitHelper.sol";
+import {SafeCallback} from "../base/SafeCallback.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/src/types/PoolId.sol";
+
+abstract contract BaseV4Quoter is SafeCallback {
+    using SqrtPriceLimitHelper for uint160;
+    using QuoterRevert for *;
+    using PoolIdLibrary for PoolId;
+
+    error NotEnoughLiquidity(PoolId poolId);
+    error NotSelf();
+    error UnexpectedCallSuccess();
+
+    constructor(IPoolManager _poolManager) SafeCallback(_poolManager) {}
+
+    /// @dev Only this address may call this function. Used to mimic internal functions, using an
+    /// external call to catch and parse revert reasons
+    modifier selfOnly() {
+        if (msg.sender != address(this)) revert NotSelf();
+        _;
+    }
+
+    function _unlockCallback(bytes calldata data) internal override returns (bytes memory) {
+        (bool success, bytes memory returnData) = address(this).call(data);
+        // Every quote path gathers a quote, and then reverts either with QuoteSwap(quoteAmount) or alternative error
+        if (success) revert UnexpectedCallSuccess();
```

Comment:
Just starting to look through this but my first thought here.. is I'm wondering if its even possible that this returns true? Like could you write a test that hits this revert?

Reply:
yeah I dont think so - but its such a tiny check for a lens contract to do that feels worth catching?